File: Source/Misc/RTR.cs
using System;
using WCSharp.Api;
using static WCSharp.Api.Common;

public class RTR
{
    private const float RTR_INTERVAL = 0.001f;
    private const float ITEM_PICKUP_RADIUS = 48;

    private Kitty kitty;
    private timer RTRTimer;
    private bool enabled;

    private trigger ClickTrigger { get; set; }
    private trigger WidgetTrigger { get; set; }

    private float targetX = 0;
    private float targetY = 0;
    private bool hasTarget = false;
    private string lastUnitAnimation;
    public float? absoluteMoveSpeed = null;

    public RTR(Kitty kitty)
    {
        this.kitty = kitty;
        RTRTimer = timer.Create();
        enabled = false;
        RegisterClickEvent();
    }

    public bool IsEnabled()
    {
        return enabled;
    }

    public void StartRTR()
    {
        enabled = true;
        kitty.CanEarnAwards = false;
        ResumeRTR();
    }

    public void ResumeRTR()
    {
        if (!enabled)
        {
            return;
        }

        ClickTrigger.Enable();
        WidgetTrigger.Enable();

        RTRTimer.Start(RTR_INTERVAL, true, ErrorHandler.Wrap(UpdateRTR));
    }

    public void PauseRTR()
    {
        ClickTrigger.Disable();
        WidgetTrigger.Disable();
        RTRTimer.Pause();
        hasTarget = false;
        if (this.lastUnitAnimation != "stand")
        {
            this.lastUnitAnimation = "stand";
            kitty.Unit.SetAnimation(0);
        }
    }

    public void StopRTR()
    {
        enabled = false;
        kitty.CanEarnAwards = true;
        this.PauseRTR();
    }

    private void UpdateRTR()
    {
        if (!hasTarget)
        {
            if (this.lastUnitAnimation != "stand")
            {
                this.lastUnitAnimation = "stand";
                kitty.Unit.SetAnimation(0);
            }
            return;
        }

        float currentX = kitty.Unit.X;
        float currentY = kitty.Unit.Y;

        float distanceToTarget = WCSharp.Shared.Util.DistanceBetweenPoints(currentX, currentY, targetX, targetY);

        if (distanceToTarget < 10)
        {
            hasTarget = false;
            if (this.lastUnitAnimation != "stand")
            {
                this.lastUnitAnimation = "stand";
                kitty.Unit.SetAnimation(0);
            }
            return;
        }

        if (this.lastUnitAnimation != "walk")
        {
            this.lastUnitAnimation = "walk";
            kitty.Unit.SetAnimation(6);
        }

        float moveSpeed = this.absoluteMoveSpeed ?? GetUnitMoveSpeed(kitty.Unit);
        float movePerTick = moveSpeed * RTR_INTERVAL;

        float angle = Atan2(targetY - currentY, targetX - currentX);
        SetUnitFacing(kitty.Unit, angle * Blizzard.bj_RADTODEG);

        float newX = currentX + (movePerTick * Cos(angle));
        float newY = currentY + (movePerTick * Sin(angle));

        if (IsTerrainPathable(newX, currentY, PATHING_TYPE_WALKABILITY))
        {
            newX = currentX;
        }

        if (IsTerrainPathable(currentX, newY, PATHING_TYPE_WALKABILITY))
        {
            newY = currentY;
        }

        kitty.Unit.SetPosition(newX, newY);
        ItemPickup();
    }

    private void RegisterClickEvent()
    {
        ClickTrigger = trigger.Create();
        ClickTrigger.RegisterUnitEvent(kitty.Unit, unitevent.IssuedPointOrder);
        ClickTrigger.AddAction(() => HandleClick(true));

        WidgetTrigger = trigger.Create();
        WidgetTrigger.RegisterUnitEvent(kitty.Unit, unitevent.IssuedTargetOrder);
        WidgetTrigger.AddAction(() => HandleClick(false));

        ClickTrigger.Disable();
        WidgetTrigger.Disable();
    }

    private void HandleClick(bool isToLocation)
    {
        if (!IsEnabled()) return;

        IssueImmediateOrder(@event.Unit, "stop");

        if (isToLocation)
        {
            targetX = GetOrderPointX();
            targetY = GetOrderPointY();
        }
        else
        {
            var target = GetOrderTarget();
            targetX = GetWidgetX(target);
            targetY = GetWidgetY(target);
        }

        hasTarget = true;
    }

    private void ItemPickup()
    {
        if (!enabled) return;

        var kibbleList = ItemSpatialGrid.GetNearbyKibbles(kitty.Unit.X, kitty.Unit.Y);
        var itemList = ItemSpatialGrid.GetNearbyItems(kitty.Unit.X, kitty.Unit.Y);

        if (kibbleList != null && kibbleList.Count > 0)
        {
            for (int i = 0; i < kibbleList.Count; i++)
            {
                var k = kibbleList[i];
                if (k == null) continue;
                kitty.Unit.AddItem(k.Item);
            }
        }

        if (itemList != null && itemList.Count > 0)
        {
            for (int i = 0; i < itemList.Count; i++)
            {
                var item = itemList[i];
                if (item == null) continue;
                if (item.IsOwned) continue;
                kitty.Unit.AddItem(item);
                break;
            }
        }
    }
}

File: src/Misc/RTR.ts
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { ItemSpatialGrid } from 'src/Game/Items/ItemSpatialGrid'
import { distanceBetweenXYPoints } from 'src/Utility/Utility'
import { getTriggerUnit } from 'src/Utility/w3tsUtils'
import { Timer, Trigger } from 'w3ts'

export class RTR {
    private RTR_INTERVAL = 0.001
    private ITEM_PICKUP_RADIUS = 48

    private kitty: Kitty
    private RTRTimer: Timer
    private enabled = false

    private ClickTrigger: Trigger
    private WidgetTrigger: Trigger

    private targetX = 0
    private targetY = 0
    private hasTarget = false
    private lastUnitAnimation: string
    public absoluteMoveSpeed = 0

    public constructor(kitty: Kitty) {
        this.kitty = kitty
        this.RTRTimer = Timer.create()
        this.enabled = false
        this.RegisterClickEvent()
    }

    public IsEnabled(): boolean {
        return this.enabled
    }

    public StartRTR = () => {
        this.enabled = true
        this.kitty.CanEarnAwards = false
        this.ResumeRTR()
    }

    public ResumeRTR = () => {
        if (!this.enabled) {
            return
        }

        this.ClickTrigger.enabled = true
        this.WidgetTrigger.enabled = true

        this.RTRTimer.start(this.RTR_INTERVAL, true, this.UpdateRTR)
    }

    public PauseRTR = () => {
        if (!this.enabled) {
            return
        }
        this.ClickTrigger.enabled = false
        this.WidgetTrigger.enabled = false
        this.RTRTimer.pause()
        this.hasTarget = false
        if (this.lastUnitAnimation !== 'stand') {
            this.lastUnitAnimation = 'stand'
            this.kitty.Unit.setAnimation(0)
        }
    }

    public StopRTR = () => {
        this.enabled = false
        // kitty.CanEarnAwards = true;
        this.PauseRTR()
    }

    private UpdateRTR = () => {
        if (!this.hasTarget) {
            if (this.lastUnitAnimation !== 'stand') {
                this.lastUnitAnimation = 'stand'
                this.kitty.Unit.setAnimation(0)
            }
            return
        }

        const currentX = this.kitty.Unit.x
        const currentY = this.kitty.Unit.y

        const distanceToTarget = distanceBetweenXYPoints(currentX, currentY, this.targetX, this.targetY)

        if (distanceToTarget < 10) {
            this.hasTarget = false
            if (this.lastUnitAnimation !== 'stand') {
                this.lastUnitAnimation = 'stand'
                this.kitty.Unit.setAnimation(0)
            }
            return
        }

        if (this.lastUnitAnimation !== 'walk') {
            this.lastUnitAnimation = 'walk'
            this.kitty.Unit.setAnimation(6)
        }

        const moveSpeed = this.absoluteMoveSpeed ?? this.kitty.Unit.moveSpeed
        const movePerTick = moveSpeed * this.RTR_INTERVAL

        const angle = Atan2(this.targetY - currentY, this.targetX - currentX)
        this.kitty.Unit.facing = angle * bj_RADTODEG

        let newX = currentX + movePerTick * Cos(angle)
        let newY = currentY + movePerTick * Sin(angle)

        if (IsTerrainPathable(newX, currentY, PATHING_TYPE_WALKABILITY)) {
            newX = currentX
        }

        if (IsTerrainPathable(currentX, newY, PATHING_TYPE_WALKABILITY)) {
            newY = currentY
        }

        this.kitty.Unit.setPathing(false)
        this.kitty.Unit.setPosition(newX, newY)
        this.kitty.Unit.setPathing(true)

        this.ItemPickup()
    }

    private RegisterClickEvent = () => {
        this.ClickTrigger = Trigger.create()!
        this.ClickTrigger.registerUnitEvent(this.kitty.Unit, EVENT_UNIT_ISSUED_POINT_ORDER)
        this.ClickTrigger.addAction(() => this.HandleClick(true))

        this.WidgetTrigger = Trigger.create()!
        this.WidgetTrigger.registerUnitEvent(this.kitty.Unit, EVENT_UNIT_ISSUED_TARGET_ORDER)
        this.WidgetTrigger.addAction(() => this.HandleClick(false))

        this.ClickTrigger.enabled = false
        this.WidgetTrigger.enabled = false
    }

    private HandleClick = (isToLocation: boolean) => {
        if (!this.IsEnabled()) return

        getTriggerUnit().issueImmediateOrder('stop')

        if (isToLocation) {
            this.targetX = GetOrderPointX()
            this.targetY = GetOrderPointY()
        } else {
            const target = GetOrderTarget()!
            this.targetX = GetWidgetX(target)
            this.targetY = GetWidgetY(target)
        }

        this.hasTarget = true
    }

    private ItemPickup = () => {
        if (!this.enabled) return
        ItemSpatialGrid.KittyItemPickup(this.kitty)
    }
}
